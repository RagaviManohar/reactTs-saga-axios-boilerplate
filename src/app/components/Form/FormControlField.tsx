/**
 *
 */

import isEmpty from 'lodash/isEmpty';
import { FormControl, FormHelperText } from '@mui/material';

import {
  LabelProps,
  ErrorProps,
} from 'app/components/Form/types/FormControlField.types';

export const FormControlField = ({ children }) => {
  return children;
};

FormControlField.FormControl = ({ children, ...props }) => {
  return <FormControl {...props}>{children}</FormControl>;
};

FormControlField.Label = ({
  name,
  displayName,
  required,
  className,
  ...props
}: LabelProps) => {
  return (
    <label className={className} htmlFor={name} {...props}>
      {displayName} {required ? '*' : ''}
    </label>
  );
};

FormControlField.Error = ({ message, ...props }: ErrorProps) => {
  if (!isEmpty(message))
    return (
      <FormHelperText error {...props}>
        {message}
      </FormHelperText>
    );

  return <></>;
};

export default FormControlField;
