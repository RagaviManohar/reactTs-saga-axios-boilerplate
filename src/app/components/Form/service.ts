import { Validate } from 'react-hook-form';

import { PATTERNS } from 'utils/constants';

type validateStringType = {
  value: string;
  message: string;
};

export const validateString = ({ value, message }: validateStringType) => {
  if (value.trim().length === 0) {
    return message;
  }
};

export const emailValidation =
  (message: string): Validate<string> =>
  (value: string) => {
    const isValid = PATTERNS.email.test(value);
    if (!isValid) {
      return message;
    }
  };

interface lengthValidationType {
  emptyStringErrorMessage: string;
  minErrorMessage: string;
  maxErrorMessage: string;
  minLimit?: Number;
  maxLimit?: Number;
}

// Empty string validation
export const lengthValidation =
  ({
    emptyStringErrorMessage,
    minErrorMessage,
    maxErrorMessage,
    minLimit = 1,
    maxLimit = 100,
  }: lengthValidationType): Validate<string> =>
  (value: string) => {
    if (value?.length > 0) {
      validateString({ value, message: emptyStringErrorMessage });

      if (value.trim().length < minLimit) {
        return minErrorMessage;
      }
      if (value?.trim().length > maxLimit) {
        return maxErrorMessage;
      }
    }
  };
