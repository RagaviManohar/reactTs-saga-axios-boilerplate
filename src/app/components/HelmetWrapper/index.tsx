import { Helmet } from 'react-helmet-async';

interface HelmetWrapperProps {
  title: string;
  description?: string;
}

export const HelmetWrapper = ({ title, description }: HelmetWrapperProps) => (
  <Helmet>
    <title>{title}</title>
    {description && <meta name="description" content={description} />}
  </Helmet>
);

export default HelmetWrapper;
