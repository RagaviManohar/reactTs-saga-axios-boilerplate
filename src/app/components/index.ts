//
// Form
//
export { default as Form } from './Form';
//
// HelmetWrapper
//
export { default as HelmetWrapper } from './HelmetWrapper';
//
// PrivateRoute
//
export { default as PrivateRoute } from './PrivateRoute';
