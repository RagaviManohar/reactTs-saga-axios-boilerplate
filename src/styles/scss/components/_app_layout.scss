.app-dashboard {
  &-page {
    height: 100vh;
  }
  &-inside {
    padding-top: 64px;
    @include sizing(100%, calc(100vh - 64px));
  }
  &-sidebar {
    max-width: $sidebar-width;
    @include sizing(100%, calc(100vh - 64px));
  }
  &-content {
    @include sizing(100%, calc(100vh - 64px));
    overflow: auto;
  }
}

//Not found
.notfound-page-content {
  height: calc(100vh - 64px);
  .notfound-img-section {
    img {
      max-width: 100%;
    }
  }
}

//Responsive design
@include small-device-below {
  .notfound-page-content {
    height: calc(100vh - 124px);
  }
  .app-dashboard {
    &-content {
      @include sizing(100%, calc(100vh - 124px));
      overflow: auto;
      &.mobile-footer-hide{
        @include sizing(100%, calc(100vh - 64px) !important);
      }
    }
  }
}
