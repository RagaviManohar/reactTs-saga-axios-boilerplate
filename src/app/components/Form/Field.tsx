/**
 * Field
 */

import { TextField, Switch } from '@mui/material';

import { generateField, typeOfFields } from 'app/components/Form/GenerateField';

const Field = {
  Input: generateField(TextField),
  Toggle: generateField(Switch, typeOfFields.toggle),
};

export default Field;
