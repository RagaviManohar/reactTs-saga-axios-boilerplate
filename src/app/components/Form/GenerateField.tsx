/**
 * generateField
 */

import { forwardRef } from 'react';

// Lodash
import uniqueId from 'lodash/uniqueId';
import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';

import { Controller, useFormContext } from 'react-hook-form';

import FormControlField from 'app/components/Form/FormControlField';

import { FieldComponentType } from 'app/components/Form/types/Field.types';

export const typeOfFields = {
  default: 'default',
  control: 'control',
  toggle: 'toggle',
  checkbox: 'checkbox',
};

export const generateField = (
  FormComponent,
  typeOfFormControl = typeOfFields.default,
) => {
  // To create an unique id with the parameter as prefix
  const fieldId = uniqueId('form-field-');

  const ModifiedFormComponent = forwardRef(
    (
      {
        rules,
        name,
        formControlProps,
        labelProps,
        renderInput,
        ...props
      }: FieldComponentType,
      ref: any,
    ) => {
      const {
        control,
        register,
        formState: { errors },
      } = useFormContext();

      const helperText = get(errors, `${name}.message`, '');

      const getComponent = () => {
        let errorProps: object = {
          error: !isEmpty(helperText),
        };

        // Toggle doesn't send error Property.
        if (
          [typeOfFields.toggle, typeOfFields.checkbox].includes(
            typeOfFormControl,
          )
        ) {
          errorProps = {};
        }

        if (
          [typeOfFields.control, typeOfFields.checkbox].includes(
            typeOfFormControl,
          )
        ) {
          return (
            <>
              <Controller
                render={({ field }) => (
                  <FormComponent
                    id={fieldId}
                    control={control}
                    inputProps={{ 'aria-label': name }}
                    {...field}
                    {...props}
                    ref={ref}
                    {...errorProps}
                  />
                )}
                control={control}
                name={name}
                rules={rules}
                {...props}
              />

              <FormControlField.Error message={helperText.toString()} />
            </>
          );
        }

        return (
          <>
            <FormComponent
              id={fieldId}
              inputProps={{ 'aria-label': name, autoComplete: 'off' }}
              {...register(name, { ...rules })}
              {...props}
              {...errorProps}
            />

            <FormControlField.Error message={helperText.toString()} />
          </>
        );
      };

      return (
        <>
          <FormControlField.FormControl {...formControlProps}>
            <FormControlField.Label name={fieldId} {...labelProps} />

            {getComponent()}
          </FormControlField.FormControl>
        </>
      );
    },
  );

  ModifiedFormComponent.displayName = 'FieldComponent';

  return ModifiedFormComponent;
};

export default generateField;
