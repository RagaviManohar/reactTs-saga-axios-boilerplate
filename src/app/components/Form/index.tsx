/**
 * Form
 */

import { memo } from 'react';
import { FormProvider } from 'react-hook-form';

import { FormType } from 'app/components/Form/types/Form.types';

export const Form = ({
  formName,
  onSubmit,
  onError,
  children,
  form,
  ...props
}: FormType) => {
  const { handleSubmit } = form;

  return (
    <FormProvider {...form}>
      <form
        onSubmit={handleSubmit(onSubmit, onError)}
        aria-label={formName}
        {...props}
      >
        {children}
      </form>
    </FormProvider>
  );
};

export default memo(Form);

export { default as Field } from './Field';
