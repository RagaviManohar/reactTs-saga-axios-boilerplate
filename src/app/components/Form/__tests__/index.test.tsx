/**
 * Tests for Form
 */

import { render, cleanup, reactHookForm } from 'jest.setup';

import { Form, Field } from 'app/components/Form';

afterEach(cleanup);

const mockFormName = 'testForm';
const mockOnSubmit = jest.fn();

const Component = () => {
  const form = reactHookForm.useForm({});
  return (
    <Form formName={mockFormName} onSubmit={mockOnSubmit} form={form}>
      <Field.Input name="firstName" />

      <button type="submit">Save</button>
    </Form>
  );
};

const renderComponent = () => {
  return render(<Component />);
};

describe('<Form/>', () => {
  it('expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    renderComponent();
    expect(spy).not.toHaveBeenCalled();
  });

  it('should render and match the snapshot', () => {
    const container = renderComponent();
    expect(container.baseElement).toMatchSnapshot();
  });
});
