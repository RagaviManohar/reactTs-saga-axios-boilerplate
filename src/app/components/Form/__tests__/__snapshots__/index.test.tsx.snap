// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form/> should render and match the snapshot 1`] = `
<body>
  <div>
    <form
      aria-label="testForm"
    >
      <div
        class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
      >
        <label
          for="form-field-1"
        >
           
        </label>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              aria-label="firstName"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="form-field-1"
              name="firstName"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</body>
`;
